File: crates/netcfgd-sys/src/rule.rs
//! Policy routing rules: the `ip rule` table.
//!
//! A rule says *which routing table to consult*, before any table is consulted.
//! That is what makes source-based routing, VRFs and multi-uplink policy work,
//! and it is why the model carries the field set `ip rule` exposes and no more
//! (`netcfgd_model::RoutingRule`).
//!
//! **Ownership comes from `FRA_PROTOCOL`**, exactly as it does for routes.
//! Decision 0002 stamps 110 on everything netcfgd installs and refuses to
//! delete anything not carrying it; rules get the same treatment from the same
//! constant. Without it a rule would be indistinguishable from one an operator
//! added by hand, and reconciliation would either remove theirs or never remove
//! its own.
//!
//! `FRA_PROTOCOL` arrived in Linux 4.17. On anything older the attribute is
//! ignored on the way in and absent on the way out, so every rule reads as
//! unowned and netcfgd installs but never removes. That is the safe direction,
//! and it is stated rather than discovered.

use crate::socket::Netlink;
use crate::wire::{self, flags, AttrBuf};
use std::io;
use std::net::IpAddr;

/// `RTM_NEWRULE`, `RTM_DELRULE`, `RTM_GETRULE`.
const RTM_NEWRULE: u16 = 32;
const RTM_DELRULE: u16 = 33;
const RTM_GETRULE: u16 = 34;

/// `FRA_*`, the rule attributes.
const FRA_DST: u16 = 1;
const FRA_SRC: u16 = 2;
const FRA_IIFNAME: u16 = 3;
const FRA_PRIORITY: u16 = 6;
const FRA_FWMARK: u16 = 10;
const FRA_SUPPRESS_PREFIXLEN: u16 = 14;
const FRA_TABLE: u16 = 15;
const FRA_FWMASK: u16 = 16;
const FRA_OIFNAME: u16 = 17;
const FRA_L3MDEV: u16 = 19;
const FRA_PROTOCOL: u16 = 21;

/// `FR_ACT_*`.
pub const FR_ACT_TO_TBL: u8 = 1;
/// Drop silently.
pub const FR_ACT_BLACKHOLE: u8 = 6;
/// Drop, and say so.
pub const FR_ACT_UNREACHABLE: u8 = 7;
/// Drop, administratively.
pub const FR_ACT_PROHIBIT: u8 = 8;

/// `FIB_RULE_INVERT`.
const FIB_RULE_INVERT: u32 = 0x0000_0002;

/// `RT_TABLE_UNSPEC`, which is what the header carries when the real table id
/// is in `FRA_TABLE` because it does not fit in a byte.
const RT_TABLE_UNSPEC: u8 = 0;

/// Length of `struct fib_rule_hdr`.
const FIB_RULE_HDR_LEN: usize = 12;

/// One rule, as it goes out and comes back.
///
/// The same type both ways, because a rule netcfgd installs and a rule it reads
/// have to be comparable field by field -- a separate "spec" and "record" pair
/// is two places for the comparison to drift.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct RuleRecord {
	/// `AF_INET` or `AF_INET6`.
	pub family: u8,
	/// Consulted in ascending order.
	pub priority: u32,
	/// Which table to look up, for [`FR_ACT_TO_TBL`].
	pub table: u32,
	/// What to do on a match.
	pub action: u8,
	/// Source selector, as address and prefix length.
	pub from: Option<(IpAddr, u8)>,
	/// Destination selector.
	pub to: Option<(IpAddr, u8)>,
	/// Incoming interface selector.
	pub iif: Option<String>,
	/// Outgoing interface selector.
	pub oif: Option<String>,
	/// Firewall mark selector.
	pub fwmark: Option<u32>,
	/// Mask applied before comparing the mark.
	pub fwmask: Option<u32>,
	/// Ignore routes shorter than this.
	pub suppress_prefixlength: Option<u32>,
	/// Match packets belonging to an l3mdev master.
	pub l3mdev: bool,
	/// Invert the selectors.
	pub invert: bool,
	/// `FRA_PROTOCOL`. 110 is netcfgd's, per decision 0002.
	pub protocol: u8,
}

/// `struct fib_rule_hdr`.
///
/// Byte-for-byte the same shape as `rtmsg`, and deliberately not that type:
/// bytes five, six and seven are `res1`, `res2` and `action` here against
/// `protocol`, `scope` and `type` there. Reusing `RtMsg` would compile and put
/// the action where the kernel reads a route type.
fn header(rule: &RuleRecord) -> Vec<u8> {
	let mut out = Vec::with_capacity(FIB_RULE_HDR_LEN);
	out.push(rule.family);
	out.push(rule.to.map_or(0, |(_, prefix)| prefix));
	out.push(rule.from.map_or(0, |(_, prefix)| prefix));
	out.push(0); // tos
			  // Only tables that fit in a byte go here; anything larger travels in
			  // `FRA_TABLE` and this stays unspecified. Truncating instead would send
			  // table 1000 as table 232.
	out.push(u8::try_from(rule.table).unwrap_or(RT_TABLE_UNSPEC));
	out.push(0); // res1
	out.push(0); // res2
	out.push(rule.action);
	out.extend_from_slice(&if rule.invert { FIB_RULE_INVERT } else { 0 }.to_ne_bytes());
	out
}

/// The attributes for one rule.
fn attributes(rule: &RuleRecord) -> AttrBuf {
	let mut attrs = AttrBuf::new();
	attrs.push_u32(FRA_PRIORITY, rule.priority);
	if rule.table != 0 {
		attrs.push_u32(FRA_TABLE, rule.table);
	}
	if let Some((address, _)) = rule.from {
		attrs.push_ip(FRA_SRC, address);
	}
	if let Some((address, _)) = rule.to {
		attrs.push_ip(FRA_DST, address);
	}
	if let Some(name) = &rule.iif {
		attrs.push_str(FRA_IIFNAME, name);
	}
	if let Some(name) = &rule.oif {
		attrs.push_str(FRA_OIFNAME, name);
	}
	if let Some(mark) = rule.fwmark {
		attrs.push_u32(FRA_FWMARK, mark);
	}
	if let Some(mask) = rule.fwmask {
		attrs.push_u32(FRA_FWMASK, mask);
	}
	if let Some(length) = rule.suppress_prefixlength {
		attrs.push_u32(FRA_SUPPRESS_PREFIXLEN, length);
	}
	if rule.l3mdev {
		attrs.push_u8(FRA_L3MDEV, 1);
	}
	attrs.push_u8(FRA_PROTOCOL, rule.protocol);
	attrs
}

impl Netlink {
	/// Every policy routing rule, both families.
	///
	/// # Errors
	///
	/// Returns the errno the kernel replied with.
	pub fn rules(&mut self) -> io::Result<Vec<RuleRecord>> {
		// An all-zero header: family `AF_UNSPEC` asks for both families in one
		// dump, which is what `ip rule show` does.
		let body = vec![0_u8; FIB_RULE_HDR_LEN];
		let replies = self.request(
			RTM_GETRULE,
			flags::NLM_F_REQUEST | flags::NLM_F_DUMP,
			&body,
			&AttrBuf::new(),
		)?;
		Ok(replies
			.iter()
			.filter_map(|payload| decode(payload))
			.collect())
	}

	/// Install a rule.
	///
	/// # Errors
	///
	/// Returns the errno the kernel replied with. `EEXIST` means a rule with
	/// this priority and family is already there.
	pub fn add_rule(&mut self, rule: &RuleRecord) -> io::Result<()> {
		self.request(
			RTM_NEWRULE,
			flags::NLM_F_REQUEST | flags::NLM_F_ACK | flags::NLM_F_CREATE | flags::NLM_F_EXCL,
			&header(rule),
			&attributes(rule),
		)?;
		Ok(())
	}

	/// Remove a rule.
	///
	/// The request carries `FRA_PROTOCOL`, and the kernel matches a delete
	/// against every attribute it is given -- so a delete asking for protocol
	/// 110 cannot match a rule that does not carry it. That is a second layer
	/// under the planner's ownership check, at the kernel rather than in
	/// netcfgd, and it means even a planner bug cannot remove somebody else's
	/// rule.
	///
	/// Verified rather than assumed: a rule installed with protocol 0 survives
	/// a delete sent with 110, and goes away when sent with 0.
	///
	/// # Errors
	///
	/// Returns the errno the kernel replied with, except `ENOENT`, which means
	/// it was already gone -- which is also what a non-matching delete looks
	/// like.
	pub fn del_rule(&mut self, rule: &RuleRecord) -> io::Result<()> {
		match self.request(
			RTM_DELRULE,
			flags::NLM_F_REQUEST | flags::NLM_F_ACK,
			&header(rule),
			&attributes(rule),
		) {
			Ok(_) => Ok(()),
			Err(error) if error.raw_os_error() == Some(libc::ENOENT) => Ok(()),
			Err(error) => Err(error),
		}
	}
}

/// One rule out of a dump.
fn decode(payload: &[u8]) -> Option<RuleRecord> {
	if payload.len() < FIB_RULE_HDR_LEN {
		return None;
	}
	let attrs = wire::Attrs::new(payload.get(FIB_RULE_HDR_LEN..)?);
	let family = *payload.first()?;
	let dst_len = *payload.get(1)?;
	let src_len = *payload.get(2)?;
	let table_byte = *payload.get(4)?;
	let action = *payload.get(7)?;
	let rule_flags = u32::from_ne_bytes(payload.get(8..12)?.try_into().ok()?);

	// The suppressor is dumped as all-ones when unset, which is not a prefix
	// length. Reading it literally would make every ordinary rule look as
	// though it suppressed prefixes shorter than four billion.
	let suppress = attrs
		.get(FRA_SUPPRESS_PREFIXLEN)
		.and_then(|attr| attr.u32())
		.filter(|value| *value != u32::MAX);

	Some(RuleRecord {
		family,
		priority: attrs
			.get(FRA_PRIORITY)
			.and_then(|attr| attr.u32())
			.unwrap_or(0),
		table: attrs
			.get(FRA_TABLE)
			.and_then(|attr| attr.u32())
			.unwrap_or_else(|| u32::from(table_byte)),
		action,
		from: attrs
			.get(FRA_SRC)
			.and_then(|attr| attr.ip())
			.map(|ip| (ip, src_len)),
		to: attrs
			.get(FRA_DST)
			.and_then(|attr| attr.ip())
			.map(|ip| (ip, dst_len)),
		iif: attrs.get(FRA_IIFNAME).and_then(|attr| attr.string()),
		oif: attrs.get(FRA_OIFNAME).and_then(|attr| attr.string()),
		fwmark: attrs.get(FRA_FWMARK).and_then(|attr| attr.u32()),
		fwmask: attrs.get(FRA_FWMASK).and_then(|attr| attr.u32()),
		suppress_prefixlength: suppress,
		l3mdev: attrs.get(FRA_L3MDEV).and_then(|attr| attr.u8()) == Some(1),
		invert: rule_flags & FIB_RULE_INVERT != 0,
		protocol: attrs
			.get(FRA_PROTOCOL)
			.and_then(|attr| attr.u8())
			.unwrap_or(0),
	})
}

File: c/src/sys/rule.c
/*
 * rule.c -- one `RTM_GETRULE` reply, as a rule.
 *
 * No descriptor is opened, read or closed in this file, which is dump.c's
 * property and for the same reason: everything here takes bytes the caller
 * already has, so every case is testable against a request this port itself
 * encoded, with no kernel, no privilege and no policy routing on the machine.
 *
 * The attribute helpers below are this file's own rather than dump.c's. They
 * are the same four lines -- find an attribute, and let "not there" be absence
 * while a malformed area refuses the record -- and dump.c's are `static` in a
 * file this worker did not write. Sharing them means changing that file, which
 * is not a change a rule decoder gets to make; the alternative was to widen
 * `wire.h`, which is landed. Said here so that whoever reconciles the two later
 * knows this was a choice rather than an oversight.
 */
#include "ncfg/rule.h"

#include <string.h>

#include <linux/fib_rules.h>

/* The length this file decodes is the kernel's own, not a second opinion. */
_Static_assert(sizeof(struct fib_rule_hdr) == NCFG_RULE_HDR_LEN,
    "a rule header is twelve bytes and this one is not");

/*
 * Find an attribute, distinguishing "not there" from "the area is nonsense".
 *
 * Returns 0 only for the second, with `err` set: a missing `FRA_PROTOCOL` is a
 * kernel older than 4.17, while a malformed area is a message nothing should be
 * decoded out of.
 */
static int area_find(const ncfg_wire_attrs_t *area, uint16_t kind, ncfg_wire_attr_t *out,
    int *found, char *err, size_t err_size)
{
	ncfg_wire_step_t step = ncfg_wire_attrs_find(area, kind, out, err, err_size);

	*found = (step == NCFG_WIRE_OK);
	return step != NCFG_WIRE_BAD;
}

static int area_u8(const ncfg_wire_attrs_t *area, uint16_t kind, uint8_t *value, int *found,
    char *err, size_t err_size)
{
	ncfg_wire_attr_t attr;
	int              present = 0;

	*found = 0;
	if (!area_find(area, kind, &attr, &present, err, err_size)) {
		return 0;
	}
	/* An attribute of the wrong width is absent, deliberately: `wire.c`
	 * refuses a value of the wrong size rather than truncating it, and
	 * reading one with the wrong accessor would return a number that is
	 * merely wrong. */
	if (!present || !ncfg_wire_attr_u8(&attr, value, NULL, 0)) {
		return 1;
	}
	*found = 1;
	return 1;
}

static int area_u32(const ncfg_wire_attrs_t *area, uint16_t kind, uint32_t *value, int *found,
    char *err, size_t err_size)
{
	ncfg_wire_attr_t attr;
	int              present = 0;

	*found = 0;
	if (!area_find(area, kind, &attr, &present, err, err_size)) {
		return 0;
	}
	if (!present || !ncfg_wire_attr_u32(&attr, value, NULL, 0)) {
		return 1;
	}
	*found = 1;
	return 1;
}

static int area_string(const ncfg_wire_attrs_t *area, uint16_t kind, char *out, size_t out_size,
    int *found, char *err, size_t err_size)
{
	ncfg_wire_attr_t attr;
	int              present = 0;

	*found = 0;
	if (!area_find(area, kind, &attr, &present, err, err_size)) {
		return 0;
	}
	/* A name that would not fit is absent rather than truncated: a
	 * truncated interface name is a name, just somebody else's. */
	if (!present || !ncfg_wire_attr_string(&attr, out, out_size, NULL, 0)) {
		return 1;
	}
	*found = 1;
	return 1;
}

static int area_ip(const ncfg_wire_attrs_t *area, uint16_t kind, ncfg_wire_ip_t *value,
    int *found, char *err, size_t err_size)
{
	ncfg_wire_attr_t attr;
	int              present = 0;

	*found = 0;
	if (!area_find(area, kind, &attr, &present, err, err_size)) {
		return 0;
	}
	if (!present || !ncfg_wire_attr_ip(&attr, value, NULL, 0)) {
		return 1;
	}
	*found = 1;
	return 1;
}

void ncfg_dump_rule_request(ncfg_buf_t *body, ncfg_buf_t *attrs)
{
	uint8_t header[NCFG_RULE_HDR_LEN];

	(void)attrs;
	/* An all-zero header: family `AF_UNSPEC` asks for both families in one
	 * dump. Written as bytes rather than through a struct so that a
	 * compiler's padding rules are not part of what goes on the wire, which
	 * is what ops_route.c does on the way out. */
	memset(header, 0, sizeof(header));
	ncfg_buf_add(body, header, sizeof(header));
}

int ncfg_dump_rule(const void *payload, size_t length, ncfg_rule_record_t *out,
    char *err, size_t err_size)
{
	const uint8_t    *bytes = (const uint8_t *)payload;
	ncfg_wire_attrs_t area;
	uint32_t          flags = 0;
	uint32_t          table = 0;
	uint8_t           table_byte;
	uint8_t           l3mdev = 0;
	int               found = 0;

	if (!out) {
		ncfg_error_set(err, err_size, "a rule record needs somewhere to go");
		return 0;
	}
	memset(out, 0, sizeof(*out));
	if (!bytes || length < NCFG_RULE_HDR_LEN) {
		ncfg_error_set(err, err_size,
		    "a rule message is at least %u bytes and this one is %zu",
		    (unsigned)NCFG_RULE_HDR_LEN, bytes ? length : (size_t)0);
		return 0;
	}
	out->family = bytes[0];
	out->to_len = bytes[1];
	out->from_len = bytes[2];
	/* bytes[3] is `tos`, which netcfgd does not carry: it is deprecated in
	 * the kernel and `ip rule` has not offered it for years. */
	table_byte = bytes[4];
	/* bytes[5] and bytes[6] are `res1` and `res2`. The action is seventh,
	 * which is where a route message keeps its type -- see rule.h. */
	out->action = bytes[7];
	memcpy(&flags, bytes + 8, sizeof(flags));
	out->invert = (flags & FIB_RULE_INVERT) != 0;

	ncfg_wire_attrs_start(&area, bytes + NCFG_RULE_HDR_LEN, length - NCFG_RULE_HDR_LEN);

	if (!area_u32(&area, FRA_PRIORITY, &out->priority, &found, err, err_size)) {
		return 0;
	}
	/* `FRA_TABLE` carries the table id when it does not fit the byte-wide
	 * field in the header, which is every table above 255. Reading only the
	 * byte reports table 1000 as table 232, and 232 is a real table
	 * belonging to somebody else. */
	if (!area_u32(&area, FRA_TABLE, &table, &found, err, err_size)) {
		return 0;
	}
	out->table = found ? table : (uint32_t)table_byte;

	if (!area_ip(&area, FRA_SRC, &out->from, &found, err, err_size)) {
		return 0;
	}
	/* A prefix length means nothing without the selector it belongs to, so
	 * it is kept only where the address arrived -- which is the reading side
	 * of what ops_route.c does when it sends one. */
	if (!found) {
		out->from_len = 0;
	}
	if (!area_ip(&area, FRA_DST, &out->to, &found, err, err_size)) {
		return 0;
	}
	if (!found) {
		out->to_len = 0;
	}

	if (!area_string(&area, FRA_IIFNAME, out->iif, sizeof(out->iif), &out->has_iif,
	    err, err_size)) {
		return 0;
	}
	if (!area_string(&area, FRA_OIFNAME, out->oif, sizeof(out->oif), &out->has_oif,
	    err, err_size)) {
		return 0;
	}
	if (!area_u32(&area, FRA_FWMARK, &out->fwmark, &out->has_fwmark, err, err_size)) {
		return 0;
	}
	if (!area_u32(&area, FRA_FWMASK, &out->fwmask, &out->has_fwmask, err, err_size)) {
		return 0;
	}
	if (!area_u32(&area, FRA_SUPPRESS_PREFIXLEN, &out->suppress_prefixlength,
	    &out->has_suppress_prefixlength, err, err_size)) {
		return 0;
	}
	/* The suppressor is dumped as all-ones when unset, which is not a prefix
	 * length. Reading it literally would make every ordinary rule look as
	 * though it suppressed prefixes shorter than four billion -- and a plan
	 * comparing that against a document that asked for nothing would rewrite
	 * every rule on the machine, for ever. */
	if (out->has_suppress_prefixlength && out->suppress_prefixlength == UINT32_MAX) {
		out->has_suppress_prefixlength = 0;
		out->suppress_prefixlength = 0;
	}
	if (!area_u8(&area, FRA_L3MDEV, &l3mdev, &found, err, err_size)) {
		return 0;
	}
	out->l3mdev = found && l3mdev == 1;
	if (!area_u8(&area, FRA_PROTOCOL, &out->protocol, &found, err, err_size)) {
		return 0;
	}
	if (!found) {
		/* Not netcfgd's, which is what a kernel older than 4.17 makes
		 * every rule on the machine look like. Installs go on working
		 * and removals stop, which is the safe direction. */
		out->protocol = 0;
	}
	return 1;
}

void ncfg_rule_record_spec(const ncfg_rule_record_t *record, ncfg_ops_rule_t *out)
{
	if (!out) {
		return;
	}
	memset(out, 0, sizeof(*out));
	if (!record) {
		return;
	}
	out->family = record->family;
	out->priority = record->priority;
	out->table = record->table;
	out->action = record->action;
	out->from = record->from;
	out->from_len = record->from_len;
	out->to = record->to;
	out->to_len = record->to_len;
	/* Borrowed, not copied. See rule.h: the spec must not outlive the
	 * record, which is the same arrangement it has with a parsed document on
	 * the other side. */
	out->iif = record->has_iif ? record->iif : NULL;
	out->oif = record->has_oif ? record->oif : NULL;
	out->fwmark.has = record->has_fwmark;
	out->fwmark.value = (int64_t)record->fwmark;
	out->fwmask.has = record->has_fwmask;
	out->fwmask.value = (int64_t)record->fwmask;
	out->suppress_prefixlength.has = record->has_suppress_prefixlength;
	out->suppress_prefixlength.value = (int64_t)record->suppress_prefixlength;
	out->l3mdev = record->l3mdev;
	out->invert = record->invert;
	/* Present, always. A record states what the kernel reported, including
	 * that it reported nothing -- while an absent protocol on the way *out*
	 * means netcfgd's own 110. Leaving this absent would turn a rule
	 * somebody else installed into one this port would claim. */
	out->protocol.has = 1;
	out->protocol.value = record->protocol;
}
